src/native/custom_key_op_handler.c

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "custom_key_op_handler.h"
+
+static void s_aws_custom_key_op_handler_perform_operation(
+    struct aws_custom_key_op_handler *key_op_handler,
+    struct aws_tls_key_operation *operation) {
+
+    struct aws_jni_custom_key_op_handler *op_handler = (struct aws_jni_custom_key_op_handler *)key_op_handler->impl;
+    AWS_FATAL_ASSERT(op_handler != NULL);
+
+    /* Get the Java ENV */
+    JNIEnv *env = aws_jni_acquire_thread_env(op_handler->jvm);
+    AWS_FATAL_ASSERT(env != NULL);
+
+    jbyteArray jni_input_data = NULL;
+    jobject jni_operation = NULL;
+    bool success = false;
+    bool exception_occured = false;
+
+    if (operation == NULL) {
+        goto clean_up;
+    }
+
+    /* Create DirectByteBuffer */
+    struct aws_byte_cursor input_data = aws_tls_key_operation_get_input(operation);
+    jni_input_data = aws_jni_byte_array_from_cursor(env, &input_data);
+    if (jni_input_data == NULL) {
+        aws_jni_check_and_clear_exception(env);
+        goto clean_up;
+    }
+
+    /* Create TlsKeyOperation */
+    jni_operation = (*env)->NewObject(
+        env,
+        tls_key_operation_properties.cls,
+        tls_key_operation_properties.constructor,
+        (jlong)(intptr_t)operation,
+        jni_input_data,
+        (jint)aws_tls_key_operation_get_type(operation),
+        (jint)aws_tls_key_operation_get_signature_algorithm(operation),
+        (jint)aws_tls_key_operation_get_digest_algorithm(operation));
+    if (jni_operation == NULL) {
+        aws_jni_check_and_clear_exception(env);
+        goto clean_up;
+    }
+
+    /**
+     * Invoke TlsKeyOperationHandler.performOperation() through the invokePerformOperation
+     * function. This function will also catch any exceptions and clear the operation
+     * with an exception should it occur.
+     */
+    (*env)->CallVoidMethod(
+        env, op_handler->jni_custom_key_op, tls_key_operation_handler_properties.perform_operation_id, jni_operation);
```

Comment:
my idea:
DON'T call the virtual function directly.
INSTEAD call some internal helper function, that calls the virtual function in a try/catch block, and then does op.completeExceptionally() entirely out in Java.

and our Java TlsKeyOperation class is protected against double-complete, so we're all good

and this function is probably impossible to fail, so we can just FATAL_ASSERT or something